### PR TITLE
feat: update auth guard code

### PR DESCRIPTION
**Pull request by Harish-osmosys**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](https://github.com/OsmosysSoftware/osmo-x/blob/main/CONTRIBUTING.md#-submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [x] I have performed unit testing for the new feature added or updated to ensure the new features added are working as expected.
- [ ] I have added/updated test cases to the [test suite](https://github.com/OsmosysSoftware/osmo-x/blob/main/apps/api/OsmoX.postman_collection.json) as applicable.
- [ ] I have performed preliminary testing using the [test suite](https://github.com/OsmosysSoftware/osmo-x/blob/main/apps/api/OsmoX.postman_collection.json) to ensure that any existing features are not impacted and any new features are working as expected as a whole.
- [ ] I have added/updated the required [api docs](https://github.com/OsmosysSoftware/osmo-x/tree/main/apps/api/docs) as applicable.
- [ ] I have added/updated the `.env.example` file with the required values as applicable.

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login endpoint`)
- [x] Description has been added
- [x] Related changes have been added (optional)
- [x] Screenshots have been added (optional)
- [ ] Query request and response examples have been added (as applicable, in case added or updated)
- [ ] Documentation changes have been listed (as applicable)
- [ ] Test suite/unit testing output is added (as applicable)
- [ ] Pending actions have been added (optional)
- [ ] Any other additional notes have been added (optional)

### Additional Information

- [x] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [x] Assignee(s) and reviewer(s) have been added (optional)

_Note: Reviewer should ensure that the checklist and description have been populated and followed correctly, and the PR should be merged only after resolving all conversations and verifying that CI checks pass._

---

**Description:**
This Pr update the logic for authorization of the endpoints earlier we were using the server api key which was wrong 
Have update it to use the jwt token for login and other endpoint like - creating provider , creating application 

**Related changes:**

- Add new values in .env.example

**Screenshots:**

image

**Query request and response:**

NA

**Documentation changes:**

NA

**Test suite/unit testing output:**

NA

**Pending actions:**

NA 

**Additional notes:**

NA

## Release Notes

- **New Features**
	- Introduced JWT authentication capabilities, enhancing security for user authentication and authorization.
	- Added support for GraphQL authentication with the new `GqlAuthGuard`, replacing the previous API key-based guard.
	- Updated environment variables to include `JWT_SECRET` and `JWT_EXPIRES_IN` for improved JWT handling.

- **Bug Fixes**
	- Streamlined authentication logic across various modules, ensuring a more secure and efficient user validation process.

- **Refactor**
	- Removed dependencies on API key management, shifting towards a JWT-centric approach in multiple services and modules.
	- Simplified response structures in the login process, focusing solely on token management.

- **Documentation**
	- Updated environment variables in the configuration to support JWT functionality.

### CODE REVIEW

File: apps/api/src/common/guards/api-key/gql-auth.guard.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  getRequest(context: ExecutionContext) {
+    const ctx = GqlExecutionContext.create(context);
+    return ctx.getContext().req;
+  }
+}
+
+export const CurrentUser = createParamDecorator((data: unknown, context: ExecutionContext) => {
+  const ctx = GqlExecutionContext.create(context);
+  return ctx.getContext().req.user;
+});
```
[the remark this answers is not in the document]
Comment (xixas) [reply apps/api/src/common/guards/api-key/gql-auth.guard.ts:17]: No

File: apps/api/src/modules/applications/applications.module.ts [hunk 2]
```diff
@@ -3,21 +3,14 @@ import { ApplicationsService } from './applications.service';
...
-  imports: [TypeOrmModule.forFeature([Application]), ServerApiKeysModule, UsersModule],
-  providers: [
-    ApplicationsService,
-    ApplicationsResolver,
-    ServerApiKeysService,
-    UsersService,
-    Logger,
-  ],
+  imports: [TypeOrmModule.forFeature([Application]), JwtModule, UsersModule],
+  providers: [ApplicationsService, ApplicationsResolver, JwtService, UsersService, Logger],
```
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/applications/applications.module.ts:13]: we still use the ServerApiKeysService in notification creation using the rest controller

[the remark this answers is not in the document]
Comment (xixas) [reply apps/api/src/modules/applications/applications.module.ts:13]: They are both different

File: apps/api/src/modules/auth/auth.service.ts [hunk 3]
```diff
@@ -3,76 +3,53 @@ import { LoginResponse } from './dto/login-response';
...
   }
 
   async login(loginUserInput: LoginUserInput): Promise<LoginResponse> {
     try {
-      const token = await this.getUserToken(loginUserInput.username);
-      const tokenList = await this.setTokenList(loginUserInput.username);
+      const user = await this.validateUser(loginUserInput.username, loginUserInput.password);
 
+      if (!user) {
+        throw new Error('Invalid credentials. Please provide valid username and password.');
```
[the remark this answers is not in the document]
Comment (xixas) [reply apps/api/src/modules/auth/auth.service.ts:40]: @Harish-osmosys  Can update this.

File: apps/api/src/modules/auth/auth.service.ts [hunk 4]
```diff
@@ -3,76 +3,53 @@
...
+    private jwtService: JwtService,
   ) {}
 
-  async validateUser(username: string, password: string): Promise<string> {
-    const entry = await this.usersService.findByUsername(username);
+  async validateUser(username: string, password: string): Promise<User> {
+    const user = await this.usersService.findByUsername(username);
 
-    if (!entry || !entry.username) {
+    if (!user || !user.username) {
```
[the remark this answers is not in the document]
Comment (xixas) [reply apps/api/src/modules/auth/auth.service.ts:20]: Update this

File: apps/api/src/modules/auth/auth.service.ts [hunk 5]
```diff
@@ -3,76 +3,53 @@
...
+      const payload: JwtPayload = {
+        username: user.username,
+        userId: user.userId,
+        role: user.userRole,
+      };
       return {
-        token,
-        user: loginUserInput.username,
-        allKeys: tokenList,
+        token: this.jwtService.sign(payload),
```
[the remark this answers is not in the document]
Comment (xixas) [reply apps/api/src/modules/auth/auth.service.ts:49]: @Harish-osmosys  What is the current esxpiry, I remember keeping some expiry in env.

File: apps/api/src/modules/applications/applications.service.ts [hunk 6]
```diff
@@ -52,16 +52,23 @@ export class ApplicationsService extends CoreService<Application> {
...
-      // Find the related application entry
-      const applicationEntry = await this.findById(apiKeyEntry.applicationId);
-      // Find the related user entry
-      const userEntry = await this.usersService.findByUserId(applicationEntry.userId);
+      // Decode the token to get the payload (which includes the user ID)
+      const decodedToken = this.jwtService.decode(token) as { userId: number };
 
-      // Check if ADMIN
+      if (!decodedToken || !decodedToken.userId) {
+        throw new UnauthorizedException('Invalid token');
```
[the remark this answers is not in the document]
Comment (xixas) [reply apps/api/src/modules/applications/applications.service.ts:61]: @Harish-osmosys Check this comment once.

File: apps/api/.env.example [hunk 7]
```diff
@@ -4,6 +4,8 @@ REQUEST_MAX_SIZE= # Max Size of Request, Default is 50mb
 CLEANUP_IDLE_RESOURCES=false # Cleans up idle queues if inactive for the specified duration, default false
 IDLE_TIMEOUT=30m # How long the queue should be idle before being considered for deletion, default 30m
 CLEANUP_INTERVAL=7d # Frequency for running the cleanup, use formats from https://github.com/vercel/ms, default 7d
+JWT_SECRET=your-strong-secret-key-here # A strong secret key for signing your JWTs. This should be a long, random string.
+JWT_EXPIRES_IN=1h #Common formats are '60s', '10m', '1h', '1d', etc
```
Comment (xixas) [inline apps/api/.env.example:8]: Keep it a month

File: apps/api/src/modules/auth/strategies/jwt.strategy.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+      ignoreExpiration: false,
+      secretOrKey: configService.getOrThrow('JWT_SECRET'),
+    });
+  }
+
+  async validate(payload: JwtPayload): Promise<JwtPayload> {
+    return {
+      userId: payload.userId,
+      username: payload.username,
+      role: payload.role,
```
Comment (xixas) [inline apps/api/src/modules/auth/strategies/jwt.strategy.ts:21]: Do we have roles right now?

**Review by xixas — CHANGES_REQUESTED** [review]

File: apps/api/src/modules/applications/applications.service.ts [hunk 9]
```diff
@@ -52,16 +52,23 @@
   async checkAdminUser(authHeader: Request): Promise<boolean> {
     try {
       const bearerToken = authHeader.toString();
-      const apiKeyToken = bearerToken.substring(7);
+      const token = bearerToken.substring(7);
```
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/applications/applications.service.ts]: updated in latest commit : https://github.com/OsmosysSoftware/osmo-x/pull/330/commits/46fca140b1e9844f1a6137dac10bcd7ea11a3976

File: apps/api/src/modules/applications/applications.service.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/applications/applications.service.ts]: updated as per the suggestion : https://github.com/OsmosysSoftware/osmo-x/pull/330/commits/46fca140b1e9844f1a6137dac10bcd7ea11a3976

File: apps/api/src/modules/auth/auth.service.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/auth/auth.service.ts:51]: We already have expiry setup in the auth.module.ts

File: apps/api/src/modules/auth/auth.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/auth/auth.service.ts]: Have updated as per suggestion

File: apps/api/src/modules/auth/auth.service.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/auth/auth.service.ts]: Update as per suggstion in :https://github.com/OsmosysSoftware/osmo-x/pull/330/commits/46fca140b1e9844f1a6137dac10bcd7ea11a3976

File: apps/api/src/modules/auth/strategies/jwt.strategy.ts (hunk 8 above)
Comment (Harish-osmosys) [reply apps/api/src/modules/auth/strategies/jwt.strategy.ts:21]: we have 2 role 

normal user - 0 
admin user - 1

File: apps/api/src/modules/applications/applications.service.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/applications/applications.service.ts]: "JsonWebTokenError: secret or public key must be provided",
                    "    at /home/harish-work/Osmosys/osmo-x/apps/api/node_modules/jsonwebtoken/verify.js:113:19",
                    "    at getSecret (/home/harish-work/Osmosys/osmo-x/apps/api/node_modules/jsonwebtoken/verify.js:97:14)",
                    
                    getting this error after this suggestion i even tried passing like this :
                          const decodedToken = this.jwtService.verify(token, secret);

File: apps/api/src/modules/auth/auth.service.ts [hunk 10]
```diff
@@ -2,77 +1,57 @@
...
```
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/auth/auth.service.ts:51]: As earlier this is already done in the auth.module.ts

File: apps/api/src/modules/auth/auth.service.ts [hunk 11]
```diff
@@ -1,78 +1,57 @@
...
+    if (user && valid) {
+      delete user.password;
+      return user;
     }
 
     return null;
   }
 
   async login(loginUserInput: LoginUserInput): Promise<LoginResponse> {
     try {
```
[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/auth/auth.service.ts:36]: we already handling these exceptions

File: apps/api/.env.example (hunk 7 above)
Comment (Harish-osmosys) [reply apps/api/.env.example]: done

File: apps/api/src/modules/applications/applications.service.ts [hunk 12]
```diff
@@ -52,16 +54,29 @@ export class ApplicationsService extends CoreService<Application> {
...
+
+      const token = bearerToken.slice(7).trim();
+      const secret = this.configService.getOrThrow('JWT_SECRET');
+
+      // Decode the token to get the payload (which includes the user ID)
+      const decodedToken = this.jwtService.verify(token, { secret });
+
+      if (!decodedToken || !decodedToken.userId) {
+        throw new UnauthorizedException('Invalid token');
+      }
```
[the remark this answers is not in the document]
Comment (xixas) [reply apps/api/src/modules/applications/applications.service.ts:70]: It is set

File: apps/api/src/modules/applications/applications.service.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (xixas) [reply apps/api/src/modules/applications/applications.service.ts]: How are they related?
Which change caused it?

Also add values in .env.example if not added.

[the remark this answers is not in the document]
Comment (Harish-osmosys) [reply apps/api/src/modules/applications/applications.service.ts]: so this was related to the verify funciton we also have to pass the secret  but we have to pass like this 
this.jwtService.verify(token, { secret } );

now it working fine have tested it

**Review by xixas — APPROVED** [review]

**Pull request merged**